modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionDemander.java

```diff
@@ -300,21 +292,58 @@ Runnable addAssignments(
         if ((delay == 0 || force) && assignments != null) {
             final RebalanceFuture oldFut = rebalanceFut;
 
-            final RebalanceFuture fut = new RebalanceFuture(grp, assignments, log, rebalanceId);
+            if (assignments.cancelled()) { // Pending exchange.
+                if (log.isDebugEnabled())
+                    log.debug("Rebalancing skipped due to cancelled assignments.");
+
+                return null;
+            }
+
+            if (assignments.isEmpty()) { // Nothing to rebalance.
+                if (log.isDebugEnabled())
+                    log.debug("Rebalancing skipped due to empty assignments.");
+
+                if (oldFut.isInitial())
+                    oldFut.onDone(true);
+
+                ((GridFutureAdapter)grp.preloader().syncFuture()).onDone();
 
-            if (!grp.localWalEnabled())
+                ctx.exchange().scheduleResendPartitions();
+
+                return null;
+            }
+
+            if (!force && (!oldFut.isDone() || oldFut.result()) && oldFut.compatibleWith(assignments)) {
+                commonRebalanceFuture.add(oldFut);
+
+                return null;
+            }
+
+            final RebalanceFuture fut = new RebalanceFuture(grp, assignments, log, rebalanceId, next);
+
+            if (!grp.localWalEnabled()) {
                 fut.listen(new IgniteInClosureX<IgniteInternalFuture<Boolean>>() {
                     @Override public void applyx(IgniteInternalFuture<Boolean> future) throws IgniteCheckedException {
                         if (future.get())
-                            ctx.walState().onGroupRebalanceFinished(grp.groupId(), assignments.topologyVersion());
+                            ctx.walState().onGroupRebalanceFinished(grp.groupId());
                     }
                 });
+            }
 
             if (!oldFut.isInitial())
                 oldFut.cancel();
             else
                 fut.listen(f -> oldFut.onDone(f.result()));
 
+            for (Map.Entry<ClusterNode, GridDhtPartitionDemandMessage> e : assignments.entrySet()) {
```

Comment:
A comment is required explaining why clearing was moved here.

Reply:
Added comments to that place in code.